src/script.c

```diff
@@ -0,0 +1,422 @@
+    if (!server.cluster_enabled || original_c->id == CLIENT_ID_AOF || (original_c->flags & CLIENT_MASTER)) {
+        return C_OK;
+    }
+    /* If this is a Redis Cluster node, we need to make sure the script is not
+     * trying to access non-local keys, with the exception of commands
+     * received from our master or when loading the AOF back in memory. */
+    int error_code;
+    /* Duplicate relevant flags in the script client. */
+    c->flags &= ~(CLIENT_READONLY | CLIENT_ASKING);
+    c->flags |= original_c->flags & (CLIENT_READONLY | CLIENT_ASKING);
+    if (getNodeByQuery(c, c->cmd, c->argv, c->argc, NULL, &error_code) != server.cluster->myself) {
+        if (error_code == CLUSTER_REDIR_DOWN_RO_STATE) {
+            *err = sdsnew(
+                    "Script attempted to execute a write command while the "
+                            "cluster is down and readonly");
+        } else if (error_code == CLUSTER_REDIR_DOWN_STATE) {
+            *err = sdsnew("Script attempted to execute a command while the "
+                    "cluster is down");
+        } else {
+            *err = sdsnew("Script attempted to access a non local key in a "
+                    "cluster node");
+        }
+        return C_ERR;
+    }
+    return C_OK;
+}
+
+static void scriptEmitMultiIfNeeded(scriptRunCtx *r_ctx) {
+    /* If we are using single commands replication, we need to wrap what
+     * we propagate into a MULTI/EXEC block, so that it will be atomic like
+     * a Lua script in the context of AOF and slaves. */
+    client *c = r_ctx->c;
+    if (!(r_ctx->flags & SCRIPT_EVAL_REPLICATION)
+         && !(r_ctx->flags & SCRIPT_MULTI_EMMITED)
+         && !(r_ctx->original_client->flags & CLIENT_MULTI)
+         && (r_ctx->flags & SCRIPT_WRITE_DIRTY)
+         && ((r_ctx->repl_flags & PROPAGATE_AOF)
+             || (r_ctx->repl_flags & PROPAGATE_REPL)))
+    {
+        execCommandPropagateMulti(r_ctx->original_client->db->id);
+        r_ctx->flags |= SCRIPT_MULTI_EMMITED;
+        /* Now we are in the MULTI context, the lua_client should be
+         * flag as CLIENT_MULTI. */
+        c->flags |= CLIENT_MULTI;
+    }
+}
+
+/* set RESP for a given r_ctx */
+int scriptSetResp(scriptRunCtx *r_ctx, int resp) {
+    if (resp != 2 && resp != 3) {
+        return C_ERR;
+    }
+
+    r_ctx->c->resp = resp;
+    return C_OK;
+}
+
+/* set Repl for a given r_ctx
+ * either: PROPAGATE_AOF | PROPAGATE_REPL*/
+int scriptSetRepl(scriptRunCtx *r_ctx, int repl) {
+    if ((repl & ~(PROPAGATE_AOF | PROPAGATE_REPL)) != 0) {
+        return C_ERR;
+    }
+    r_ctx->repl_flags = repl;
+    return C_OK;
+}
+
+/* Call a Redis command.
+ * The reply is written to the r_ctx client and it is
+ * up to the engine to take and parse.
+ * The err out variable is set only if error occurs and describe the error.
+ * If err is set on reply is written to the r_ctx client. */
+void scriptCall(scriptRunCtx *r_ctx, robj* *argv, int argc, sds *err) {
```

Comment:
Can we call r_ctx, "run_ctx", while reading this I keep thinking it must be a "redis context".